pkg/resource/deploy/source_eval.go

```diff
@@ -576,16 +576,21 @@ func getProviderReference(defaultProviders *defaultProviders, req providers.Prov
 // package with the given unparsed provider reference. If the unparsed provider reference is empty,
 // this function returns the plugin for the indicated package's default provider.
 func getProviderFromSource(
-	providers ProviderSource, defaultProviders *defaultProviders,
-	req providers.ProviderRequest, rawProviderRef string) (plugin.Provider, error) {
+	providerSource ProviderSource, defaultProviders *defaultProviders,
+	req providers.ProviderRequest, rawProviderRef string,
+	token tokens.ModuleMember) (plugin.Provider, error) {
 
 	providerRef, err := getProviderReference(defaultProviders, req, rawProviderRef)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("getProviderFromSource: %w", err)
+	} else if providers.IsDenyDefaultsProvider(providerRef) {
+		msg := diag.GetDefaultProviderDenied("Invoke").Message
+		return nil, fmt.Errorf(msg, req.Package(), token)
```

Comment:
Nit: Does "req.Package()" return a string containing a format tag? If so, code smell?

Either way, would it be easy to refactor & remove token as an argument to the function? It's only used in the error path and callers could insert it themselves.

Reply:
`req.Package()` does not contain a format tag, but `msg` contains 2. I'm not sure how we would refactor it out without changing the calling convention of our `diag` error messages.

Reply:
got it. seems out of scope for this PR.